db/snapshot/backup_task.cc

```diff
@@ -118,8 +133,30 @@ future<> backup_task_impl::do_backup() {
     _backup_shard = this_shard_id();
     co_await _sharded_worker.start(std::ref(_snap_ctl.db()), _table_id, std::ref(*this));
 
+    gate abort_gate;
+    auto abort_sub = _as.subscribe([&] () noexcept {
+        if (auto gh = abort_gate.try_hold()) {
+            std::ignore = _sharded_worker.invoke_on_all([] (worker& m) {
+                m.abort();
+            }).finally([gh = std::move(gh)] {});
+        }
+    });
+
+    // Errors from `start_uploading` are retrieved by `done`
+    co_await _sharded_worker.invoke_on_all([] (worker& m) {
+        return m.start_uploading();
+    });
+
     try {
-        co_await uploads_worker();
+        co_await _sharded_worker.invoke_on_all([] (worker& m) {
+            return m.done();
+        });
+
+        co_await abort_gate.close();
```

Comment:
Shouldn't abort_gate be close()-ed on exception too?

Reply:
yes, if shard upload failed for some other reasons than abort, we should close the abort gate before stopping the _sharded_worker.